docs/runtime/dotnet-metrics.md

```diff
@@ -0,0 +1,679 @@
+  - [Metric: `dotnet.jit.methods_compiled.count`](#metric-dotnetjitmethods_compiledcount)
+  - [Metric: `dotnet.jit.compilation.time`](#metric-dotnetjitcompilationtime)
+- [Thread pool](#thread-pool)
+  - [Metric: `dotnet.thread_pool.threads.count`](#metric-dotnetthread_poolthreadscount)
+  - [Metric: `dotnet.thread_pool.work_items.count`](#metric-dotnetthread_poolwork_itemscount)
+  - [Metric: `dotnet.thread_pool.queue.length`](#metric-dotnetthread_poolqueuelength)
+- [General](#general)
+  - [Metric: `dotnet.monitor.lock_contention.count`](#metric-dotnetmonitorlock_contentioncount)
+  - [Metric: `dotnet.timers.count`](#metric-dotnettimerscount)
+  - [Metric: `dotnet.assemblies.count`](#metric-dotnetassembliescount)
+  - [Metric: `dotnet.exceptions.count`](#metric-dotnetexceptionscount)
+
+<!-- tocstop -->
+
+## CLR Garbage Collection
+
+**Status**: [Experimental][DocumentStatus]
+
+**Description:** .NET Common Language Runtime (CLR) metrics relating to garbage collection, captured under the namespace `dotnet.gc.*`.
+
+### Metric: `dotnet.gc.collections.count`
+
+This metric is [recommended][MetricRecommended].
+This metric reports the same values as calling [`GC.CollectionCount(int generation)`](https://learn.microsoft.com/en-us/dotnet/api/system.gc.collectioncount).
+
+<!-- semconv metric.dotnet.gc.collections.count(metric_table) -->
+<!-- NOTE: THIS TEXT IS AUTOGENERATED. DO NOT EDIT BY HAND. -->
+<!-- see templates/registry/markdown/snippet.md.j2 -->
+<!-- prettier-ignore-start -->
+<!-- markdownlint-capture -->
+<!-- markdownlint-disable -->
+
+| Name     | Instrument Type | Unit (UCUM) | Description    | Stability |
+| -------- | --------------- | ----------- | -------------- | --------- |
+| `dotnet.gc.collections.count` | Counter | `{collection}` | Number of garbage collections that have occurred since the process has started. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+
+
+<!-- markdownlint-restore -->
+<!-- prettier-ignore-end -->
+<!-- END AUTOGENERATED TEXT -->
+<!-- endsemconv -->
+
+<!-- semconv metric.dotnet.gc.collections.count(full) -->
+<!-- NOTE: THIS TEXT IS AUTOGENERATED. DO NOT EDIT BY HAND. -->
+<!-- see templates/registry/markdown/snippet.md.j2 -->
+<!-- prettier-ignore-start -->
+<!-- markdownlint-capture -->
+<!-- markdownlint-disable -->
+
+| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
+|---|---|---|---|---|---|
+| [`dotnet.gc.generation`](/docs/attributes-registry/dotnet.md) | string | Name of the garbage collector heap generation. | `gen0`; `gen1`; `gen2`; `loh`; `poh` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+
+
+<!-- markdownlint-restore -->
+<!-- prettier-ignore-end -->
+<!-- END AUTOGENERATED TEXT -->
+<!-- endsemconv -->
+
+### Metric: `dotnet.gc.objects.size`
```

Comment:
In the current contrib implementation:

- `dotnet.gc.heap.size` includes fragmentation. On .NET 6, it uses reflection to call into a non-public method (due to a bug in .NET), but on newer runtimes, it uses `GC.GetGCMemoryInfo().GenerationInfo.[x].SizeAfterBytes`.
- `dotnet.gc.objects.size` excludes fragmentation and other GC-committed memory pools. It uses `GC.GetTotalMemory(false)`.

Reply:
so `dotnet.gc.objects.size` would be the difference between `dotnet.gc.heap.size` and `dotnet.gc.heap.fragmentation` (sum across all generation). 

If so, does it feel important enough to report (given that someone can derive it if they really need it)?

If so, can we find a more descriptive name for it? (I have a vague memory of discussing it already but can't find the comment, sorry  if I'm repeating myself).

E.g. `dotnet.gc.memory.currently_allocated` (to work nicely with  `dotnet.gc.memory.total_allocated`)?

Reply:
There is also a subtle time-of-measurement difference in there which can be important. The per genereration heap.size and fragmentation.size measurements are values computed whenever the last GC collection occurred whereas objects.size is a live measurement of the current GC heap size in between collections. 

In practice this time of measurement difference can be large. For example imagine an app that has a long-lived data structures that are 100MB in size and then it processes requests which allocate some short lived objects. On a big server with lots of cores and cache the GC might decide it will be efficient to let the app allocate 500MB of objects before doing even a gen0 collection. So the heap.size metric will consistently report around ~100MB but the object.size metric will show a saw-tooth pattern where memory climbs from 100 up to 600, then drops back down to 100 again. I believe it is the only metric in the set which reveals the intra-collection behavior.

Reply:
nice, thanks for the explanation and I agree it's great to have this metric. 

I think it'd be useful to add a note summarizing the difference (the intra-collection part).

WRT naming, what do you all think of `dotnet.gc.memory.currently_allocated` ? 
Happy to explore other naming options highlighting that it's a current measurement (as opposed to `dotnet.gc.memory.total_allocated`)

Reply:
> WRT naming, what do you all think of dotnet.gc.memory.currently_allocated ?

I'm happy with that. I assume based on the other discussion we might also change 'memory' => 'heap' ?